File: src/main/java/nl/joelchrist/shorty/shorties/endpoints/ShortiesEndpoint.java
package nl.joelchrist.shorty.shorties.endpoints;

import nl.joelchrist.shorty.exceptions.DuplicateShortyException;
import nl.joelchrist.shorty.exceptions.EntityNotFoundException;
import nl.joelchrist.shorty.shorties.domain.Shorty;
import nl.joelchrist.shorty.shorties.managers.ShortiesManager;
import org.hibernate.validator.constraints.URL;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;
import org.springframework.web.servlet.view.RedirectView;

import javax.servlet.http.HttpServletResponse;
import javax.validation.Valid;
import java.net.URI;
import java.util.logging.Logger;

@RestController
@CrossOrigin()
public class ShortiesEndpoint {
    static Logger log = Logger.getLogger(ShortiesEndpoint.class.getName());

    @Autowired
    private ShortiesManager shortiesManager;

    @RequestMapping(value = "/shorties", method = RequestMethod.POST)
    public ResponseEntity<Shorty> createShorty(@Valid @RequestBody Shorty shorty) {
        String fullUrl = shorty.getFullUrl();
        log.info(String.format("Creating shorty for full url: %s", fullUrl));
        Shorty createdShorty = shortiesManager.createShorty(fullUrl);
        URI location = getUriFromCurrentRequest();
        return ResponseEntity.created(location).body(createdShorty);
    }

    @RequestMapping(value = "/shorties/custom", method = RequestMethod.POST)
    public ResponseEntity<Shorty> createShortyWithIdentifier(@Valid @RequestBody Shorty shorty) throws DuplicateShortyException {
        log.info(String.format("Creating shorty for full url: %s", shorty.getFullUrl()));
        Shorty createdShorty = shortiesManager.createShorty(shorty);
        URI location = getUriFromCurrentRequest();
        return ResponseEntity.created(location).body(createdShorty);
    }

    @RequestMapping(value = "/{shortIdentifier}", method = RequestMethod.GET)
    public RedirectView useShorty(@PathVariable("shortIdentifier") String shortIdentifier) throws EntityNotFoundException {
        log.info(String.format("Using shorty with shortIdentifier %s", shortIdentifier));
        Shorty shorty = shortiesManager.findShorty(shortIdentifier);
        RedirectView redirectView = new RedirectView();
        redirectView.setUrl(shorty.getFullUrl());
        return redirectView;
    }

    private URI getUriFromCurrentRequest() {
        return ServletUriComponentsBuilder.fromCurrentRequest().build().toUri();
    }

}

File: src/main/kotlin/nl/joelchrist/shorty/shorties/endpoints/ShortiesEndpoint.kt
package nl.joelchrist.shorty.shorties.endpoints

import nl.joelchrist.shorty.shorties.domain.Shorty
import nl.joelchrist.shorty.shorties.domain.ShortyMetadata
import nl.joelchrist.shorty.shorties.managers.FileManager
import nl.joelchrist.shorty.shorties.managers.ShortiesManager
import nl.joelchrist.shorty.util.getLogger
import org.hibernate.validator.constraints.NotBlank
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.io.ByteArrayResource
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.servlet.support.ServletUriComponentsBuilder
import java.util.UUID
import javax.servlet.http.HttpServletResponse
import javax.validation.constraints.NotNull

@RestController
@CrossOrigin
class ShortiesEndpoint(@Autowired private val shortiesManager: ShortiesManager, @Autowired private val fileManager: FileManager) {
    companion object {
        private val log = getLogger(ShortiesEndpoint::class)
        private fun getUriFromCurrentRequest() = ServletUriComponentsBuilder.fromCurrentRequest().build().toUri()
    }

    @RequestMapping(value = ["/shorties"], method = [(RequestMethod.POST)])
    fun createShorty(@Validated @RequestBody shorty: Shorty): ResponseEntity<Shorty> =
            ResponseEntity
                    .created(getUriFromCurrentRequest())
                    .body(shorty.let(shortiesManager::create))
                    .also { log.info("Creating shorty for url ${shorty.url}") }

    @RequestMapping(value = ["/files"], method = [(RequestMethod.POST)])
    fun createFile(@Validated @NotNull @NotBlank file: MultipartFile): ResponseEntity<ShortyMetadata> =
            ResponseEntity
                    .created(getUriFromCurrentRequest())
                    .body(UUID.randomUUID().let { ShortyMetadata(fileManager.create(it, file), it.toString()) })
                    .also { log.info("Creating file ${file.originalFilename}") }

    @RequestMapping(value = ["/{identifier}"], method = [(RequestMethod.GET)])
    fun useShorty(@PathVariable(value = "identifier") identifier: String, res: HttpServletResponse): ResponseEntity<Any> =
            shortiesManager.find(identifier)
                    .let {
                        when (it.metadata) {
                            null -> {
                                res.sendRedirect(it.url)
                                log.info("Using file with identifier $identifier")
                                ResponseEntity.ok().build()
                            }
                            else -> {
                                val resource = it
                                        .let { fileManager.find(it.metadata!!.uuid) }
                                        .also { log.info("Using file with identifier $identifier") }

                                ResponseEntity
                                        .ok()
                                        .header(HttpHeaders.CONTENT_DISPOSITION, resource.contentDisposition)
                                        .contentLength(resource.getContent().size.toLong())
                                        .contentType(MediaType.parseMediaType(resource.contentType))
                                        .body(ByteArrayResource(resource.getContent()))
                            }
                        }
                    }
}